src/optimizer/deliminator.cpp

```diff
@@ -0,0 +1,208 @@
+public:
+	DeliminatorPlanUpdater() {
+	}
+	//! Update the plan after a DelimGet has been removed
+	void VisitOperator(LogicalOperator &op) override;
+	void VisitExpression(unique_ptr<Expression> *expression) override;
+	//! Whether the operator has one or more children of type DELIM_GET
+	bool HasChildDelimGet(LogicalOperator &op);
+
+	expression_map_t<Expression *> expr_map;
+	column_binding_map_t<bool> projection_map;
+	unique_ptr<LogicalOperator> temp_ptr;
+};
+
+void DeliminatorPlanUpdater::VisitOperator(LogicalOperator &op) {
+	VisitOperatorChildren(op);
+	VisitOperatorExpressions(op);
+	// now check if this is a delim join that can be removed
+	if (op.type == LogicalOperatorType::LOGICAL_DELIM_JOIN && !HasChildDelimGet(op)) {
+		auto &delim_join = (LogicalDelimJoin &)op;
+		auto decs = &delim_join.duplicate_eliminated_columns;
+		for (auto &cond : delim_join.conditions) {
+			if (cond.comparison != ExpressionType::COMPARE_EQUAL) {
+				continue;
+			}
+			auto &colref = (BoundColumnRefExpression &)*cond.right;
+			if (projection_map.find(colref.binding) != projection_map.end()) {
+				// value on the right is a projection of removed DelimGet
+				for (idx_t i = 0; i < decs->size(); i++) {
+					if (decs->at(i)->Equals(cond.left.get())) {
+						// the value on the left no longer needs to be a duplicate-eliminated column
+						decs->erase(decs->begin() + i);
+						break;
+					}
+				}
+				// whether we applied an IS NOT NULL filter
+				cond.null_values_are_equal = true; // projection_map[colref.binding];
+			}
+		}
+		// change type if there are no more duplicate-eliminated columns
+		if (decs->empty()) {
+			delim_join.type = LogicalOperatorType::LOGICAL_COMPARISON_JOIN;
+		}
+	}
+}
+
+void DeliminatorPlanUpdater::VisitExpression(unique_ptr<Expression> *expression) {
+	if (expr_map.find(expression->get()) != expr_map.end()) {
+		*expression = expr_map[expression->get()]->Copy();
+	} else {
+		VisitExpressionChildren(**expression);
+	}
+}
+
+bool DeliminatorPlanUpdater::HasChildDelimGet(LogicalOperator &op) {
+	if (op.type == LogicalOperatorType::LOGICAL_DELIM_GET) {
+		return true;
+	}
+	for (auto &child : op.children) {
+		if (HasChildDelimGet(*child)) {
+			return true;
+		}
+	}
+	return false;
+}
+
+unique_ptr<LogicalOperator> Deliminator::Optimize(unique_ptr<LogicalOperator> op) {
+	vector<unique_ptr<LogicalOperator> *> candidates;
+	FindCandidates(&op, candidates);
+
+	for (auto candidate : candidates) {
+		DeliminatorPlanUpdater updater;
+		if (RemoveCandidate(candidate, updater)) {
+			updater.VisitOperator(*op);
+		}
+	}
+	return op;
+}
+
+void Deliminator::FindCandidates(unique_ptr<LogicalOperator> *op_ptr,
+                                 vector<unique_ptr<LogicalOperator> *> &candidates) {
+	auto op = op_ptr->get();
+	// search children before adding, so the deepest candidates get added first
+	for (auto &child : op->children) {
+		FindCandidates(&child, candidates);
+	}
+	if ( // Projection/Aggregate
```

Comment:
I would prefer to break up this huge if statement into several blocks using early outs, e.g.

```c
// search for a projection or aggregate
if (op->type != LogicalOperatorType::LOGICAL_PROJECTION &&
    op->type != LogicalOperatorType::LOGICAL_AGGREGATE_AND_GROUP_BY) {
    return;
}
// followed by a join
if (op->children[0]->type != LogicalOperatorType::LOGICAL_COMPARISON_JOIN) {
    return;
}
...

```

This keeps the same logic but is more readable than a ton of nested conditions (in my opinion).

Reply:
Agreed. I split it up and it is much more readable now